### ISSUE TITLE
Cannot validate certificate hostname without SNI

### ISSUE THREAD

**Issue by anka-213** [mitmproxy/mitmproxy#1846]
I get the following error message (example.com replaced with actual website) and can't figure out why and what to do about it:

<< Cannot establish TLS with webmail.chalmers.se:443 (sni: None): TlsException('Cannot validate certificate hostname without SNI',)

##### Steps to reproduce the problem:

1. mitmdump -vw filename
2. Connect (add as proxy) with android phone and install certificate (using mitmproxy.it)
3. Connect to some server

##### Any other comments? What have you tried so far?

It seems to work for some https-sites but not others. Examples that cause the errors are webmail.chalmers.se and m.handelsbanken.se.

##### System information

```
Mitmproxy version: 0.18.2
Python version: 3.5.2
Platform: Linux-4.7.2-1-ARCH-x86_64-with-arch
SSL version: OpenSSL 1.0.2h  3 May 2016
Linux distro: arch  
```

**Comment by mhils** [comment]
Which client are your using? I suspect an old version of Android?

The problem is as follows: When mitmproxy connects upstream, it verifies the certificate the server presents. One part of the verification is to check if the certificate matches the hostname the client is expecting. Modern versions of TLS mandate the client to send a Server Name Indication (SNI) extension during the TLS handshake which names the expected site. However, if your client does not send a SNI extension, there's no way for us to verify the authenticity of the server's certificate as we don't know which site the client is expecting and we fail the connection in the way you described. You can work around this by disabling certificate verification with `mitmproxy --insecure` - be warned that you are subject to man-in-the-middle attacks then.

**Comment by anka-213** [comment]
Yeah, that was probably the issue. Thanks!

**Comment by GrmpCerber** [comment]
>However, if your client does not send a SNI extension, there's no way for us to verify the authenticity of the server's certificate as we don't know which site the client is expecting

@mhils  I get your point on SNI yet, if a Fully Qualified Domain Name is provided with the CONNECT, can't you use this piece of data instead of the SNI ?

**Comment by mhils** [comment]
> yet, if a Fully Qualified Domain Name is provided with the CONNECT, can't you use this piece of data instead of the SNI ?

Yes, that would be possible, but I feel the complexity of supporting that outweighs the potential benefit.

**Comment by mhofman** [comment]
I'm not sure I understand why mitmproxy tries to validate the name of certificate.
IMO, the following should work and avoid upstream man-in-the-middle attacks:
- Connect to the remote server, transmitting any SNI, ALPN, etc. received from the clienty (basically copies the Client Hello info).
- When server sends Certificate, validate the chain but ignore the subject. (completely bypass verification if `--insecure`)
- Generate MITM cert based on Cert received from server (CN, other fields). Basically the only difference should be the public/private key and of course the signature (and possibly the Issuer if not replicating the cert tree)
- Client verifies the MITM Cert, including the subject against its expectations.

That way, it would work for protocols/client that don't use SNI, in transparent / SOCKS mode, or if the client tries to validate the subject using something else than a host name (happens with TCP over TLS streams).

**Comment by hgbstu** [comment]
@mhils Do you have an opinion on what @mhofman proposed? I also can't see an argument why the upstream certificate chain should not be requested if the there is no SNI extension present in the Client Hello?

**Comment by mhils** [comment]
I think @mhofman's suggestion is reasonable for the specific use case, but not suitable for mitmproxy in general:

- Most importantly, there are cases where we don't have a server connection, but still need to generate an interception certificate. One example would be server replay, which we want to work offline in a plane. Additionally, users should be able to redirect requests in scripts even if the original target host is offline.
- In many cases the hostname the client is expecting is not in the server certificate, e.g. when mitmproxy is running in reverse proxy mode. We currently include the SNI value in the certificate, which is incompatible with @mhofman's approach.
- There are users which use mitmproxy in front of old devices with broken TLS implementations to properly secure their connections.
- We want mitmproxy to be as secure as possible by default. That includes hostname validation. 

Of course we could go and add some sophisticated "client didn't send a SNI, so we now do exactly what @mhofman proposed" logic, but I feel usefulness of that doesn't warrant the additional complexity. One example where "Generate MITM cert based on Cert received from server" has gone horribly wrong is SuperFish for example.

**Comment by ploxiln** [comment]
I have another idea to work-around this problem: a reverse-dns table in the mitmproxy config.

So, if there are a couple of IP addresses which I know ahead of time will receive requests for a particular domain name, I could put that in the mitmproxy config, like:

```yaml
assume_hostname:
    "1.2.3.4": "foo.example.com"
    "5.6.7.7": "bar.test.net"
```

Then, when mitmproxy intercepts a TLS connection to 1.2.3.4, it uses the hostname foo.example.com to check the certificate, if the client did not include TLS-SNI, and it otherwise does not yet know the hostname.

Just an idea, which would help with my use case at least.

**Comment by kirill-oficerov** [comment]
@ploxiln 
when I run with the config provided by you I get in console:
> Ignored options: {'assume_hostname': CommentedMap([('my.ip.address.here', 'my.hostname')])}
and mitmproxy is started

Mitmproxy: 3.0.3 binary
Python:    3.5.2
OpenSSL:   OpenSSL 1.1.0g  2 Nov 2017
Platform:  Linux-3.13.0-79-generic-x86_64-with-debian-jessie-sid

Any clues why the option is ignored?

**Comment by ploxiln** [comment]
It does not exist. It was an idea for a feature to be added.

**Comment by plambrechtsen** [comment]
I have a similar issue as @ploxiln where the very dumb client device I am working with doesn't include the SNI but the domain came through in the previous DNS request. I can map it to an IP address so having the concept of assume_hostname would solve my problem.

**Comment by Timmmm** [comment]
Starting `mitmproxy` with the `-k` (`--ssl-insecure`) option will disable certificate checking so this will work (though obviously it will open all of your traffic up to invalid certificates).

**Comment by MWNZ-TrusteesExecutors** [comment]
> Starting `mitmproxy` with the `-k` (`--ssl-insecure`) option will disable certificate checking so this will work (though obviously it will open all of your traffic up to invalid certificates).

The disable certificate checking doesn't fix the problem with no SNI in the header. If for example you had a machine with multiple interfaces or secondary IP addresses such as:
192.168.10.1 sitea.com
192.168.10.2 siteb.com

If you could add in a "assume hostname" configuration element as mentioned above into MITM proxy then if your client is connecting to sitea.com on 192.168.10.1 but the client didn't include the SNI as part of the client hello request to the proxy server and only validated the subject but didn't care about the certificate chain (or you could load your own CA into the client) then this would work.

**Comment by plambrechtsen** [comment]
> Starting `mitmproxy` with the `-k` (`--ssl-insecure`) option will disable certificate checking so this will work (though obviously it will open all of your traffic up to invalid certificates).

The disable certificate checking doesn't fix the problem with no SNI in the header. If for example you had a machine with multiple interfaces or secondary IP addresses such as:
192.168.10.1 sitea.com
192.168.10.2 siteb.com

If you could add in a "assume hostname" configuration element as mentioned above into MITM proxy then if your client is connecting to sitea.com on 192.168.10.1 but the client didn't include the SNI as part of the client hello request to the proxy server and only validated the subject but didn't care about the certificate chain (or you could load your own CA into the client) then this would work.

**Comment by Timmmm** [comment]
> The disable certificate checking doesn't fix the problem with no SNI in the header.

I can't claim to understand SNI but I got the error listed above, and adding `--ssl-insecure` fixed it for me. I think in my case the client was connecting by IP address, like https://10.11.12.13/` or whatever, and it also ignores certificate validation errors.

Also, I found that the `mitmweb` has an option to turn on `--ssl-insecure` temporarily in the web interface which is nice. There's a slight bug in that if you pass `--ssl-insecure` on the command line, the "Don't validate certificates" checkbox is initially unchecked.
